convert2rhel/applock.py

```diff
@@ -129,31 +147,23 @@ def try_to_lock(self, _recursive=False):
         application as locked, since it is probably the result of
         manual meddling, intentional or otherwise.
 
-        :keyword _recursive: True if we are being called recursively
-                             and should not try to clean up the lockfile
-                             again.
         :raises ApplicationLockedError: the application is locked
         """
         if self._try_create():
-            self._locked = True
             return
-        if _recursive:
-            raise ApplicationLockedError("Cannot lock %s" % self._name)
-
-        with open(self._pidfile, "r") as f:
-            file_contents = f.read()
-        try:
-            pid = int(file_contents.rstrip())
-        except ValueError:
-            raise ApplicationLockedError("Lock file %s is corrupt" % self._pidfile)
-
+        pid = self._read_pidfile()
+        if pid == self._pid:
+            return
         if self._pid_exists(pid):
             raise ApplicationLockedError("%s locked by process %d" % (self._pidfile, pid))
         # The lock file was created by a process that has exited;
         # remove it and try again.
         loggerinst.info("Cleaning up lock held by exited process %d." % pid)
         os.unlink(self._pidfile)
```

Comment:
There is a race condition here too, where something else may have already `unlinked` the file so we need to catch the exception.

Also.... I think the race could also mean that self._pidfile might not contain the same pid as we checked with `_pid_exists` (meaning between us reading the pid and `unlinking` the file, another process has unlinked the original and created their new one in its place.) Is there some way to fix that with some combination of counting links to the lockfile, adding one new link to the file, and checking whether the count of links matches how many we think it should at each step of the way?  (And if it doesn't match, having some sort of `time.sleep(random)` before retrying?)  (I'm not sure if there is a way or not.  I think there might be but a quick internet search didn't turn up a recipe so I might be overly optimisitc.  If not, it's probably better to err on the side of telling the user to check whether the lock is stale rather than overwriting the lock).